### ISSUE TITLE
Support checks for nullable-mixed references

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#1118]
```php
function foo($bar) : void {
    if (rand(0, 1)) {
        $bar = null;
    }
  
    $bar->bat();
}
```

also

```php
class A {}
function foo($bar) : void {
    if (rand(0, 1)) {
        $bar = new A();
    }
  
    $bar->bat();
}
```

Expected: Psalm should warn about a possible null reference
Actual: Psalm emits no warnings
https://getpsalm.org/r/31e9204e41
